loyalty_outlier_analysis.ipynb

```diff
@@ -0,0 +1,476 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Outlier Separation\n",
+    "\n",
+    "Outliers are the ones where the target value is below -30. Most target values are clustered around 0, but then we have that bunch of values around -33, with the same value. And apparently predicting those correctly is crucial for good performance in the competition."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "train_and_validation_df['outlier'] = np.where(train_and_validation_df['target'] < -30, 'Outlier', 'Normal')\n",
+    "train_and_validation_df.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Separate the outliers into their own table"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "outlier_df = train_and_validation_df.loc[train_and_validation_df['outlier'] == 'Outlier'].copy()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "outlier_df.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Trying to see if there are any differences in the correlations between the full dataset and the outlier dataset. Practically everything looks exactly the same."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plt.figure(figsize = (14, 10))\n",
+    "sns.heatmap(train_and_validation_df.corr(), vmin=-1, vmax=1, cmap='PiYG')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plt.figure(figsize = (14, 10))\n",
+    "sns.heatmap(outlier_df.corr(), vmin=-1, vmax=1, cmap='PiYG')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Trying a Decision Tree\n",
+    "\n",
+    "Checking if a decision tree can be used to separate the outliers from the rest. Measure all of accuracy (proportion of correct predictions in all predictions), precision (proportion of true positives in predicted positives), and recall (proportion of predicted positives in actual positives)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from sklearn import tree\n",
+    "from sklearn.tree import DecisionTreeClassifier\n",
+    "from sklearn.metrics import accuracy_score, precision_score, recall_score, classification_report, confusion_matrix\n",
+    "from sklearn.model_selection import train_test_split\n",
+    "\n",
+    "def draw_decision_tree(clf, feature_names):\n",
+    "    dot_data = tree.export_graphviz(clf, out_file=None, feature_names=feature_names,\n",
```

Comment:
NB. This requires graphviz installed via homebrew – it's not enough just to install the pip package.